optimistix/_solver/gradient_methods.py

```diff
@@ -170,19 +172,31 @@ def step(
         f_eval, lin_fn, aux_eval = jax.linearize(
             lambda _y: fn(_y, args), state.y_eval, has_aux=True
         )
+
+        if self.search._needs_grad_at_y_eval:
+            grad = lin_to_grad(lin_fn, state.y_eval, autodiff_mode)
+            f_eval_info = FunctionInfo.EvalGrad(f_eval, grad)
+        else:
+            f_eval_info = FunctionInfo.Eval(f_eval)
```

Comment:
What if we need some other flavor of `FunctionInfo` later? I think picking a boolean flag here may paint us into a corner in the future. Perhaps a more general approach is required.